### PR TITLE
:bug: Fix merging of runtime.Object

### PR DESCRIPTION
**Pull request by MatiasFrank**
We now properly handle merging of child lists which contain a 'mergeKey'
Essentially, we just make sure to adhere to JSON Merge Path (https://datatracker.ietf.org/doc/html/rfc7386)

### CODE REVIEW

File: pkg/obj/merge.go [hunk 1]
```diff
@@ -11,42 +11,49 @@ import (
...
+		// 	}
+		// }
+		// where we merge 'obj's by merging their 'list' fields on 'name'. Having two obj's of
+		// source := obj{list: [{name: "n1", value: "v1"}, {name: "n2", value: "v2"}]
+		// dest := obj{list: [{name: "n1", value: "vv"}, {name: "n3", value: ""}]
+		// We expect
+		// result == obj{list: [{name: "n1", value: "vv"}, {name: "n3", value: ""}, {name: "n2", value: "v2"}]
+		// but got (note 'value' of the second element)
+		// result == obj{list: [{name: "n1", value: "vv"}, {name: "n3", value: "v2"}, {name: "n2", value: "v2"}]
+		gvk := dest.GetObjectKind().GroupVersionKind()
```
Comment (andersjohnsen) [inline pkg/obj/merge.go:51]: I have poor success with relying on GVK being set. Can er perhaps do:

```
var res T
_, _, err := serializer.Decode(out, nil, &res)
```
?

File: pkg/obj/merge.go [hunk 2]
```diff
@@ -11,42 +11,49 @@ import (
...
+	if err := serializer.Encode(dest, &dstB); err != nil {
+		return empty, fmt.Errorf("could not encode destination obj: %w", err)
 	}
 
-	out, err := strategicpatch.StrategicMergePatch(dstB.Bytes(), srcB.Bytes(), dst)
+	out, err := strategicpatch.StrategicMergePatch(dstB.Bytes(), srcB.Bytes(), dest)
 	if err == nil {
-		if _, _, err := m.serializer.Decode(out, nil, dst); err != nil {
-			return err
+		// Previously, we discarded the output and used 'dst' as 'into'. This introduced a bug in the following scenario:
```
Comment (andersjohnsen) [inline pkg/obj/merge.go:36]: This is more of a PR comment than a code comment (referring to previous implementations). Can we move this to the PR comment and add a comment here about this case generally being handled?

Comment (andersjohnsen) [reply pkg/obj/merge.go:36]: Perhaps even move it to the `Merge` godoc?

**Review by andersjohnsen — APPROVED** [review]

File: pkg/obj/merge.go (hunk 1 above)
Comment (MatiasFrank) [reply pkg/obj/merge.go:55]: We can do that now. Previously, we used `Merge` in a context where we didn't know the type of the objects, but that I've changed.

**Comment by MatiasFrank** [comment]
Comment to `Merge` changes and how we use `serializer.Decode`
Previously, we discarded the output and used `object` as the `into` argument to `Decode`. This introduced a bug in the following scenario:
Consider
```
type obj Struct {
	list []struct{
		name string
		value string
	}
}
```
where we merge 'obj's by merging their 'list' fields on 'name'. Having two obj's of
```
source := obj{list: [{name: "n1", value: "v1"}, {name: "n2", value: "v2"}]
dest := obj{list: [{name: "n1", value: "vv"}, {name: "n3", value: ""}]
```
We expect
```
result == obj{list: [{name: "n1", value: "vv"}, {name: "n3", value: ""}, {name: "n2", value: "v2"}]
```
but got (note 'value' of the second element)
```
result == obj{list: [{name: "n1", value: "vv"}, {name: "n3", value: "v2"}, {name: "n2", value: "v2"}]
```

**Pull request merged**